psl-core/src/main/java/org/linqs/psl/reasoner/dcd/DCDReasoner.java

```diff
@@ -59,82 +60,89 @@ public double optimize(TermStore baseTermStore) {
 
         termStore.initForOptimization();
 
-        // This must be called after the term store has to correct variable capacity.
-        // A reallocation can cause this array to become out-of-date.
-        float[] variableValues = termStore.getVariableValues();
-
-        double objective = -1.0;
+        long termCount = 0;
+        double objective = Double.POSITIVE_INFINITY;
         double oldObjective = Double.POSITIVE_INFINITY;
+        float[] oldVariableValues = null;
 
         if (printInitialObj && log.isTraceEnabled()) {
-            objective = computeObjective(termStore, variableValues);
-            log.trace("Iteration {} -- Objective: {}, Iteration Time: {}, Total Optimiztion Time: {}", 0, objective, 0, 0);
+            objective = computeObjective(termStore);
+            log.trace("Iteration {} -- Objective: {}, Mean Movement: {}, Iteration Time: {}, Total Optimiztion Time: {}", 0, objective, 0.0f, 0, 0);
         }
 
-        int iteration = 1;
         long totalTime = 0;
-        while (true) {
+        boolean converged = false;
+        for (int iteration = 1; iteration < (maxIterations * budget) && !converged; iteration++) {
             long start = System.currentTimeMillis();
 
+            termCount = 0;
+            objective = 0;
             for (DCDObjectiveTerm term : termStore) {
-                term.minimize(truncateEveryStep, variableValues);
+                if (oldVariableValues != null) {
```

Comment:
Is oldVariableValues only null in the initial step? If so, maybe initialize it to the first term, rather then having this if statement every check.

Reply:
oldVariableValues is only null in the initial step yes. But If we are performing streaming inference then there is no way to know if the oldVariableValues array has enough capacity for the term.evaluate() call, so initializing to the first term will not work in many situations.